projects/@shared/lib/validate-data.js

```diff
@@ -1,30 +1,64 @@
+import getImageDimensions from './get-image-dimensions';
+
 // Docs: https://joi.dev/api/?v=17.4.0#anyvalidatevalue-options
 const JOI_VALIDATION_OPTIONS = {
   abortEarly: false,
   convert: false,
 };
 
-const validateData = (data, schema) => {
-  if (!data.length) {
-    return { errors: 'No data to validate' };
-  }
+/**
+ * Transforms the data array into an object that Joi can validate.
+ *
+ * @param {Array} data
+ */
+const transformData = data => {
+  const array = data
+    .map(async item => {
+      const { type, image } = item;
+      /*
+        Because Joi Extensions cannot be async (yet): https://github.com/sideway/joi/issues/1194)
+        we get the image dimensions here and add them to the data that wil be validated.
+        ¯\_(ツ)_/¯
+      */
+      if (type === 'image') {
+        const { width, height } = await getImageDimensions(image.url);
+        const value = image.url ? { url: image.url, width, height } : image.url;
+        return { ...item, value };
+      }
+
+      return item;
+    });
 
-  // Convert incoming data into schema data object.
-  const schemaData = Object.assign({},
-    ...data.map(data => ({
-      [data.term]: data.value,
-    }))
-  );
+  return Promise.all(array)
+    .then(array => array.map(item => ({ [item.term]: item.value })))
+    .then(array => (Object.assign({}, ...array)));
+};
 
+/**
+ * Validate data with given schema and return the results.
+ *
+ * @param {Object} data
+ * @param {Object} schema
+ */
+const validateDataWithSchema = (data, schema) => {
   try {
     // Docs: https://joi.dev/api/?v=17.4.0#anyvalidate
-    const { error, warning } = schema.validate(schemaData, JOI_VALIDATION_OPTIONS);
+    const { error, warning } = schema.validate(data, JOI_VALIDATION_OPTIONS);
     // Return any errors and warnings.
     return { errors: error?.details, warnings: warning?.details };
   } catch (err) {
-    console.log(err);
     return { errors: err };
   }
 };
 
-export default validateData;
+const validate = (data, schema) => {
+  if (!data.length) {
+    return { errors: 'No data to validate' };
+  }
+
+  return transformData(data)
+    .then(data => validateDataWithSchema(data, schema))
+    .catch(err => console.log(err));
+};
```

Comment:
If there's no data, you return an `Object`, while if there is data, you return a `Promise`. Please have consistent return values. It might help to annotate this with JSDoc as well, and add the `@returns` annotation as well.

Reply:
For the same token, if there _is_ data, but something goes wrong in the `transformData` or `validateDataWithSchema` functions, it's also unclear what is returned from this function and what the `view` where `validate` is called should do.

Reply:
Changed.